### PR TITLE
Maintainer approval check

### PR DESCRIPTION
**Pull request by peternied**
### Description
_Implements proposal from https://github.com/opensearch-project/OpenSearch/issues/10613#issue-1942146755_
> Replace Codeowners to control accepting pull requests with a GitHub action check that reads from the list of maintainers any verifies they have approved the PR. Codeowners can still be used for mapping subject matter experts that want to be notified, but isn't a requirement for every maintainer to be listed on the top level.
>
> By modifying the branch protection rules 'maintainer has approved' check, could be required (only bypassable by an Admin), which would offer effectively the same security as Codeowners.

This is a new approval check that effectively doubles as the existing CODEOWNERS checks on this repository - which will allow that code owners check to be _removed_ in favor of this one.

This check pulls the list of maintainers from GitHub repository settings.  This also codifies the number of approvals needed in the repository in the event we increase or alter these requirements.

#### UX when PRs are blocked:
image

#### Example workflow:
https://github.com/peternied/OpenSearch-1/actions/runs/7026869463/job/19120342204?pr=138

#### Rollout plan:
After this PR has been merged, and been broadly accepted, a repository admin will change the required checks to include this workflow - this prevents changes from being merged even if this check isn't passing.  Then  the code owners file can be updated so maintainers can direct the kinds of changes they would like to be notified about.

### Related Issues
- Related https://github.com/opensearch-project/OpenSearch/issues/10613

### Check List
- [ ] ~New functionality includes testing.~
  - [ ] ~All tests pass~
- [ ] ~New functionality has been documented.~
  - [ ] ~New functionality has javadoc added~
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [X] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [ ] ~Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by reta** [comment]
@peternied +1 to give it a try, thank you. Just to clarify for myself, the list of collaborators automatically include ones listed in CODEOWNERS as well?

**Comment by peternied** [comment]
> @peternied +1 to give it a try, thank you. Just to clarify for myself, the list of collaborators automatically include ones listed in CODEOWNERS as well?

@reta  CODEOWNERS doesn't include permissions in the github sense - it defines a mapping between selected reviewers on PR and the file hierarchy - and it can be enforced by a branch protection setting which is currently enabled.  The permissions for users are from the repo settings which is where the list of `collaborators` comes.

After this change is merged, a maintainer could remove themselves from the CODEOWNERS entirely and still approve & merge PRs.  Personally, I plan to change my setting to watch our governance files on the root, GHAs, identity, http, and a few targeted areas of interest.  When future me is notified I know it was either because I self selected or was @mentioned. :shaking-with-excitement:

**Comment by msfroh** [comment]
I'm really excited about this change.

Before people start removing themselves from CODEOWNERS, though, I think we'll need to update (or remove?) some of the tooling that @dblock and @prudhvigodithi have to make sure that contributors and CODEOWNERS are kept in sync.

**Comment by dblock** [comment]
> I'm really excited about this change.
> 
> Before people start removing themselves from CODEOWNERS, though, I think we'll need to update (or remove?) some of the tooling that @dblock and @prudhvigodithi have to make sure that contributors and CODEOWNERS are kept in sync.

Correct, please open an issue on opensearch-build at least? cc: @bbarani

**Comment by peternied** [comment]
> I like the approach but this might be a problem in one case, what if a user raised a PR (who is not in slack) and dont know whom to tag/mention? With the existing approach of notifying every codeowner, a codeowner can tag right person (SME for the PR topic) who can review the PR. @peternied WDYT?

@prudhvigodithi The MAINTAINERS.md is always an option for finding someone, same is available with commit history. I'm not sure how often this is a problem - could you collect up data and make an issue? I'd like to address that out of band of this change.

**Comment by prudhvigodithi** [comment]
> @prudhvigodithi The MAINTAINERS.md is an option for finding someone, same with browsing PR history. I'm not sure how often this is a problem - could you collect up data and make an issue? I'd like to address that out of band of this change.

Sure Peter, we dont have to stop this change for the problem I have raised, as we never have a PR with empty reviewers with existing CODEOWNERS file setup, this would be something new and we can iterate as we learn, but is the expectation to make this the de facto standard for all repos? If so then when we run a repo audit we need to exclude `CODEOWNERS`. CC: @bbarani

**Comment by peternied** [comment]
>  is the expectation to make this the de facto standard for all repos

@prudhvigodithi Not generally - smaller repos are more manageable and this would just add overhead whereas the default tools from GitHub work well.  OpenSearch and OpenSearch Dashboards are good candidates for this change to help maintainers track incoming PRs.

However, we can always build on this change to make more improvements.

**Comment by peternied** [comment]
FYI - I was doing some additional validation on the workflow and I've run into an issue with the `peternied/approved-by-maintainers` erroring on logins with casing :/ I'll update when this has been resolved

**Comment by prudhvigodithi** [comment]
Thanks @peternied we should consider standardizing else it would deviate the standard baseline process for few repos.
@dblock @msfroh @bbarani WDYT?

**Comment by peternied** [comment]
> FYI - I was doing some additional validation on the workflow and I've run into an issue with the peternied/approved-by-maintainers erroring on logins with casing :/ I'll update when this has been resolved

This has been resolved - successful verification workflow: https://github.com/peternied/OpenSearch-1/actions/runs/7119425260/job/19421619223?pr=138

@reta @msfroh @dblock What do you think about merging this?

**Review by msfroh — APPROVED** [review]

**Review by msfroh — APPROVED** [review]

**Comment by peternied** [comment]
- All tests are passing - only github workflow file changes - https://github.com/opensearch-project/OpenSearch/issues/7396 will require a rebase, I'm merging to expedite this process.

**Comment by peternied** [comment]
@prudhvigodithi (CC @opensearch-project/admin)

Since the check is now available an running in 1.3, 2.x and main.  Could you make the following change to the branch protection on OpenSearch?

- Uncheck `Require review from Code Owners`

image

- Check `Require status checks to pass before merging`
   - Do not check or uncheck `Require branches to be up to date before merging`
   - Add `Minimum approval count` to the list of `Status checks that are required.`

image

**Comment by prudhvigodithi** [comment]
@peternied I have just updated the rule with the changes you mentioned.

**Pull request merged**
